dev/tests/api-functional/testsuite/Magento/GraphQl/Quote/Guest/SetShippingMethodsOnCartTest.php

```diff
@@ -42,100 +48,437 @@ protected function setUp()
         $this->quoteResource = $objectManager->get(QuoteResource::class);
         $this->quoteFactory = $objectManager->get(QuoteFactory::class);
         $this->quoteIdToMaskedId = $objectManager->get(QuoteIdToMaskedQuoteIdInterface::class);
+        $this->quoteAddress = $objectManager->get(QuoteAddress::class);
     }
 
+    /**
+     * @magentoApiDataFixture Magento/Checkout/_files/quote_with_virtual_product_saved.php
+     */
     public function testShippingMethodWithVirtualProduct()
     {
-        $this->markTestIncomplete('https://github.com/magento/graphql-ce/issues/422');
+        $methodCode = 'flatrate';
+        $carrierCode = 'flatrate';
+        $reservedOrderId = 'test_order_with_virtual_product_without_address';
+
+        $maskedQuoteId = $this->getMaskedQuoteIdByReversedQuoteId($reservedOrderId);
+        $quoteAddressId = $this->getQuoteAddressIdByReversedQuoteId($reservedOrderId);
+
+        $query = $this->prepareMutationQuery(
+            $maskedQuoteId,
+            $methodCode,
+            $carrierCode,
+            $quoteAddressId
+        );
+
+        self::expectException(\Exception::class);
+        self::expectExceptionMessage('Carrier with such method not found: ' . $methodCode . ', ' . $methodCode . '');
```

Comment:
Looks like we have an error in our logic because we have this method.
The error message should be like `You can't set shipping methods for virtual products`
See example
https://github.com/magento/graphql-ce/blob/2.3-develop/dev/tests/api-functional/testsuite/Magento/GraphQl/Quote/Guest/SetShippingAddressOnCartTest.php#L108

The main idea is not to having green tests but having proper functionality which covered with tests

Reply:
Changed the error message to `You can't set shipping methods for virtual products`. But because of it, this test case throwing the error.